core/src/main/java/org/dcache/nfs/util/Opaque.java

```diff
@@ -173,6 +216,19 @@ public boolean equals(Object o) {
 
             if (o instanceof OpaqueImpl) {
                 return Arrays.equals(_opaque, ((OpaqueImpl) o)._opaque);
+            } else if (o instanceof OpaqueBufferImpl) {
+                OpaqueBufferImpl other = (OpaqueBufferImpl) o;
```

Comment:
I still think this construction is semantically wrong. OpaqueBufferImpl and OpaqueImmutableImpl should not be equal, but they are.

Reply:
The general use case is that we can read an entire message into a `ByteBuffer`, and then look up what we have in map without any further allocation.

Compared to the original code of `Opaque` the "semantics" are unchanged (since Opaque was mutable).